### ISSUE TITLE
Feature Request: Add support for output in markdown format

### ISSUE THREAD

**Issue by Bohem1** [iainbrighton/PScribo#36]
We should build support for markdown output. It should be treated as a first class plugin.
Initially maybe support only Github flavor markdown but make it possible to expand so one can choose.

**Comment by Stephanevg** [comment]
I do have written this part already a bit back. Currently it is an external
function for the moment, and I am not really sure how I I can completely
integrate in pscribo. Any idea how?

**Comment by iainbrighton** [comment]
@Stephanevg I need to update the Plugin help to give you guys some guidance. If you already have part of this written, it will obviously help! Is this code hosted anywhere?

@Bohem1 @Stephanevg To write a plugin you need to write functions that will take an object (Section, Paragraph and Table etc.) are export it to the target format. Therefore, you will need to think about how you're going to render each PScribo component...

**Comment by irwins** [comment]
I was just looking through your PScribo HTML code. OutMD.ps1 & OutMD.Internal.ps1 is a start... 😀

**Comment by iainbrighton** [comment]
Yeah 😜. It might be easier if you start with the OutText plugin rather than the HTML one (this includes a whole load of style formatting that you probably won't need)?

**Comment by irwins** [comment]
[Markdown cheets](https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf) For a quick reference! 😉

**Comment by irwins** [comment]
@iainbrighton Thanks for the tip!  Isee what you mean!

**Comment by Stephanevg** [comment]
@iainbrighton  I have pasted it on a gist here -> https://gist.github.com/Stephanevg/f6409e7ab679872d69059943e5316c84

Basically, what this function does, is converting a pscribo document to MArkdown using the different ID's. There is a part I found on the internet somwhere, which I have (very) slightly modified.

 Please let me know when (and how) I can integrate this in the plugin from Scribo.

#Stéphane

**Comment by irwins** [comment]
@Stephanevg It's a lil more complicatedthan that... 😁 

I've been going through the structure (Gotta love vscode for this). The about_Pluginshelp.txt is a good start. The Matrixs tells you which features are support and which aren't or by means of options...


## Plugin support matrix

Functionality | Text |  XML  |  HTML |  Word |  CSV  |  MD
-|-|-|-|-|-|-
Document  |  X   |   X   |   X   |   X   |   X   |  X 
GlobalOption  |  X   |   X   |   X   |   X   |   X   |  X 
LineBreak  |  X   |   -   |   X   |   X   |   -   |  X
PageBreak  |  1   |   -   |   1   |   X   |   -   |  -
Paragraph  |  1   |   X   |   1   |   X   |   2   |  X
Section  |  X   |   X   |   X   |   X   |   X   |  X
Style  |  -   |   -   |   X   |   X   |   -   |  X
Table  |  3   |   3   |   X   |   X   |   2   |  3
TableStyle  |  -   |   -   |   X   |   X   |   -   |  -
Image  |  -   |   4   |   X   |   X   |   -   |  -


1. Can be implemented as LineBreaks with a plugin-specific option.
2. Each table is output in its own CSV file.
3. Table columns widths and styling are not supported.
4. Image can be embedded as a binary Base64 encoded string.

_Plugin-specific options_
_In their own about file?_

HTML: NoPageLayoutStyle [bool]: Suppresses the page layout/look-and-feel style

_Writing a plugin_

>I feel like an intern. I'll probably get there but it'll take way longer! 😄. For me it's about the collaboration experience and getting hands on experience with Git/GitHub in general...

**Comment by iainbrighton** [comment]
I probably need to do 2 things to help you here:

1. ___Write some "Writing a Plugin documentation"!___
2. Update the option handling code so there no duplication.

In the meantime @Bohem1 , @Stephanevg and @irwins (and for anyone else interested), we could look at hosting a 1 hour Skype/GoToMeeting session to take you though the text plugin and how that works. Is this of interest?

**Comment by Stephanevg** [comment]
Yeah good idea. Preferably in the evening of the week for me.

**Comment by Bohem1** [comment]
@iainbrighton a skype session is a great idea to. Preferably in the evening.

**Comment by it-praktyk** [comment]
Please inform me too - I'm interesting in direction of PScribo development because I use it [for Format-Pester](https://github.com/equelin/Format-Pester).

**Comment by ryandcoates** [comment]
Has anybody considered markdown as an input format? we have started transitioning a lot of our docs to MD to keep in source control but still require word as an end output, I was trying to tackle this pipeline with PScribo but there doesn't seem to be an easy way to take the minor format markup in a markdown file and generate the document object with the same formatting on the output side

**Comment by it-praktyk** [comment]
@ryandcoates, IMHO PSCribo isn't intended to transform existing markdown files to any other format.

PScribo uses its own DSL (a domain specific language) and mapping markdown format to it will be almost impossible. 

What about [Pandoc](https://pandoc.org/)?

**Comment by ryandcoates** [comment]
That's what I thought might be the case, sometimes I try to be a bit too creative ;)

I did try pandoc originally before someone pointed me to PScribo, it gets the data into word ok but the formatting support is fairly rudimentary, it might be the go to for a while though, I found the DSL for PScribo to be very easy to deal with when working from scratch, a lot of our scripts and tools are already taking the objects and parsing output into markdown and dumping them in source control, perhaps I could go further back in the chain and try and integrate PScribo there instead, sorry if I derailed :)

**Comment by iainbrighton** [comment]
@ryandcoates I agree with @it-praktyk here. That's not to say that someone couldn't write a MD > PScribo object conversion utility 😃. Once the MD plugin is written, you could then replace your existing scripts etc., but I'm not sure how far development has gone..

**Comment by Stephanevg** [comment]
Hey buddy @iainbrighton ! 
I have a junior colleague to which I wanted to give this task, so he has some stuff to do on the weekend (who want's to hang out in bars with friends nowadays anyways? ;)).
Is that documentation "How to write a plugin" somewhere available now? 
Perhaps you still have that recodring of that presentation you did, so I have something to get him started with.

**Comment by kilasuit** [comment]
This would be a great addition to this module if it were to surface.
